src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6357,175 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+
+                        var parameterType = ParseTypeOrVoid();
+                        types.Add(SyntaxFactory.ModifiedType(modifiers, parameterType));
+
+                        if (skipBadFunctionPointerParameterListTokens() == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+
+                        Debug.Assert(CurrentToken.Kind == SyntaxKind.CommaToken);
+                        types.AddSeparator(EatToken(SyntaxKind.CommaToken));
+                    }
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                SyntaxToken greaterThanToken;
+                if (lessThanToken.IsMissing && CurrentToken.Kind == SyntaxKind.CloseParenToken)
+                {
+                    greaterThanToken = EatTokenAsKind(SyntaxKind.GreaterThanToken);
+                }
+                else
+                {
+                    greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                }
+
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerStart()
+            => CurrentToken.Kind == SyntaxKind.DelegateKeyword && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
+
+        private bool IsPossibleCallingConvention(SyntaxToken asteriskToken)
+        {
+            if (IsPossibleFunctionPointerParameterListStart(CurrentToken))
+            {
+                return false;
+            }
+
+            // If the next token is a known function pointer calling convention, treat it as a
+            // calling convention no matter what. If it wasn't intended to be a calling convention
+            // we'd have an error anyway, and it's more likely the user intended for it to be a
+            // function pointer convention than not.
+            // PROTOTYPE(func-ptr): refactor this out into a helper method to share with binding
+            switch (CurrentToken.ValueText)
```

Comment:
>CurrentToken.ValueText [](start = 20, length = 22)

I think this change should be backed by a test that observes the difference in behavior.

Reply:
> I think this change should be backed by a test that observes the difference in behavior.

After adding a test, I've gone back to `Text` instead of `ValueText`, as we don't want to treat `@cdecl` as a calling convention, given that it's been escaped.

---
In reply to: [350896749](https://github.com/dotnet/roslyn/pull/39879#discussion_r350896749) [](ancestors = 350896749)

Reply:
> After adding a test, I've gone back to Text instead of ValueText, as we don't want to treat @cdecl as a calling convention, given that it's been escaped.

It feels like we are somewhat inconsistent:
- ```delegate* @cdecl``` - return false
- ```delegate* @cdecl <``` - return true

Perhaps we should ignore all escaped identifiers? I think it would be fine to add a prototype comment and follow up on this later.

---
In reply to: [350976752](https://github.com/dotnet/roslyn/pull/39879#discussion_r350976752) [](ancestors = 350976752,350896749)

Reply:
> It feels like we are somewhat inconsistent

I don't think that that's any more inconsistent than a non-escaped regular identifier that isn't a known convention, like `MyType`. We special case things that we know are valid calling conventions, but the convention was specifically escaped here so we treat it as any other invalid identifier.

---
In reply to: [351425780](https://github.com/dotnet/roslyn/pull/39879#discussion_r351425780) [](ancestors = 351425780,350976752,350896749)

Reply:
> I don't think that that's any more inconsistent than a non-escaped regular identifier that isn't a known convention, like MyType. We special case things that we know are valid calling conventions, but the convention was specifically escaped here so we treat it as any other invalid identifier.

Then why do we treat ```@cdecl``` as possibly valid calling convention when it is followed by ```<```. If gramar doesn't allow escaped identifiers, then we should not accept them. 

---
In reply to: [351441102](https://github.com/dotnet/roslyn/pull/39879#discussion_r351441102) [](ancestors = 351441102,351425780,350976752,350896749)